From f05864356c31d96eb824704bf415f88570699f2b Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Alex=20R=C3=B8nne=20Petersen?= <alex@alexrp.com>
Date: Sat, 18 Oct 2025 11:34:48 +0200
Subject: [PATCH 013/133] std.Thread: fix some issues in x86_64/x32 inline asm
X-Developer-Signature: v=1; a=openpgp-sha256; l=1809; i=xtex@astrafall.org;
 h=from:subject; bh=nEswiVZgSW0a8aePdVP21NTqe87hVBM3ZRf54UtHV+4=;
 b=owGbwMvMwCW2U4Ij7wZL9ETG02pJDJn7cn/VHa6dubvQeOqXleG8Bu2LXf6dnDhpDku28Ial2
 55U3f+8u2MiC4MYF4OlmCJLkWGDN6tOOr/osnJZmDmsTCBDpEUaGBgYGFgY+HIT80qNdIz0TLUN
 9QwNdYx0jBi4OAVgqi0OMfyPPCd7WH6Zjq/W9Wfn5D3q1usm1rDH/xNTrKtwOrE9J76RkWFpRjK
 fxL8nK8LKXq9WVQ1PWtKj98ZzV+IkpnWnN6/IEuUHAA==
X-Developer-Key: i=xtex@astrafall.org; a=openpgp;
 fpr=7231804B052C670F15A6771DB918086ED8045B91

Wrong syscall on x32; return exit code 0 instead of 1 on both.

ref https://github.com/ziglang/zig/issues/22189

(cherry picked from commit adcfdce6be955d7d8e3237be5c34f84478cf4636)
---
 lib/std/Thread.zig | 22 ++++++++++++++++------
 1 file changed, 16 insertions(+), 6 deletions(-)

diff --git a/lib/std/Thread.zig b/lib/std/Thread.zig
index fabe2c92c569..181ed6e25bc8 100644
--- a/lib/std/Thread.zig
+++ b/lib/std/Thread.zig
@@ -1189,12 +1189,22 @@ const LinuxThreadImpl = struct {
                     : [ptr] "r" (@intFromPtr(self.mapped.ptr)),
                       [len] "r" (self.mapped.len),
                     : .{ .memory = true }),
-                .x86_64 => asm volatile (
-                    \\  movq $11, %%rax # SYS_munmap
-                    \\  syscall
-                    \\  movq $60, %%rax # SYS_exit
-                    \\  movq $1, %%rdi
-                    \\  syscall
+                .x86_64 => asm volatile (switch (target.abi) {
+                        .gnux32, .muslx32 =>
+                        \\  movl $0x4000000b, %%eax # SYS_munmap
+                        \\  syscall
+                        \\  movl $0x4000003c, %%eax # SYS_exit
+                        \\  xor %%rdi, %%rdi
+                        \\  syscall
+                        ,
+                        else =>
+                        \\  movl $11, %%eax # SYS_munmap
+                        \\  syscall
+                        \\  movl $60, %%eax # SYS_exit
+                        \\  xor %%rdi, %%rdi
+                        \\  syscall
+                        ,
+                    }
                     :
                     : [ptr] "{rdi}" (@intFromPtr(self.mapped.ptr)),
                       [len] "{rsi}" (self.mapped.len),
-- 
2.52.0

